File: sdk/python/gust_sdk/sample.py
"""One-sample harness so ``gust test --runner http|exec|trigger`` can drive a Python agent.

End users implement a handler that runs the agent once and returns an AgentRun,
or a normal result when an existing tracer already captures the run.
gust owns sampling, optional fixtures, and the verdict.
"""

from __future__ import annotations

import json
import os
import sys
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any, Callable, Dict, IO, Optional, Union
from urllib.parse import urlparse

from gust_sdk.fixtures import FIXTURE_ENDPOINT_ENV, FixtureClient
from gust_sdk.recorder import INGEST_URL_ENV, OTEL_ENDPOINT_ENV, RunRecorder, post_run

SAMPLE_ID_ENV = "AGENTEVAL_SAMPLE_ID"
EVALUATION_ID_ENV = "GUST_EVALUATION_ID"
SCENARIO_ID_ENV = "GUST_SCENARIO_ID"
TRACE_ID_ENV = "GUST_TRACE_ID"
TRACEPARENT_ENV = "TRACEPARENT"
BAGGAGE_ENV = "BAGGAGE"
WORLD_CONTROL_ENV = "GUST_WORLD_CONTROL"
RESOURCE_ATTRS_ENV = "OTEL_RESOURCE_ATTRIBUTES"
SAMPLE_ID_ATTR = "gust.sample_id"

SampleHandler = Callable[[Dict[str, Any], FixtureClient], Union[Dict[str, Any], RunRecorder]]


def sample_context() -> Dict[str, Any]:
    """Read the Gust sample context from process environment."""
    return {
        "evaluation_id": os.environ.get(EVALUATION_ID_ENV, ""),
        "scenario_id": os.environ.get(SCENARIO_ID_ENV, ""),
        "sample_id": os.environ.get(SAMPLE_ID_ENV, ""),
        "trace_id": os.environ.get(TRACE_ID_ENV, ""),
        "traceparent": os.environ.get(TRACEPARENT_ENV, ""),
        "baggage": os.environ.get(BAGGAGE_ENV, ""),
        "world_control": os.environ.get(WORLD_CONTROL_ENV, ""),
        "tool_endpoint": os.environ.get(FIXTURE_ENDPOINT_ENV, ""),
        "input": os.environ.get("AGENTEVAL_TASK_INPUT", ""),
    }


def apply_sample_id(run: Union[Dict[str, Any], RunRecorder], sample_id: str) -> Union[Dict[str, Any], RunRecorder]:
    """Stamp ``sample_id`` onto a recorder or an AgentRun dict."""
    if not sample_id:
        return run
    if isinstance(run, RunRecorder):
        run.run_id = sample_id
        run.set_metadata("sample_id", sample_id)
        return run
    run["run_id"] = sample_id
    metadata = dict(run.get("metadata") or {})
    metadata["sample_id"] = sample_id
    run["metadata"] = metadata
    return run


def read_invoke(stdin: Optional[IO[str]] = None) -> Dict[str, Any]:
    """Read the invoke payload gust sends on stdin (exec/trigger runner)."""
    source = stdin if stdin is not None else sys.stdin
    raw = source.read()
    if not raw.strip():
        return sample_context()
    return json.loads(raw)


def _as_dict(run: Union[Dict[str, Any], RunRecorder], sample_id: str) -> Dict[str, Any]:
    stamped = apply_sample_id(run, sample_id)
    if isinstance(stamped, RunRecorder):
        return stamped.to_dict()
    return stamped


def apply_invoke_env(request: Dict[str, Any]) -> str:
    """Copy invoke fields onto the process env so OTel SDKs and export() work."""
    sample_id = str(request.get("sample_id") or os.environ.get(SAMPLE_ID_ENV, ""))
    if sample_id:
        os.environ[SAMPLE_ID_ENV] = sample_id
    for key, env_name in (
        ("evaluation_id", EVALUATION_ID_ENV),
        ("scenario_id", SCENARIO_ID_ENV),
        ("trace_id", TRACE_ID_ENV),
        ("traceparent", TRACEPARENT_ENV),
        ("baggage", BAGGAGE_ENV),
        ("world_control", WORLD_CONTROL_ENV),
    ):
        val = request.get(key)
        if val:
            os.environ[env_name] = str(val)
    endpoint = request.get("tool_endpoint") or os.environ.get(FIXTURE_ENDPOINT_ENV) or None
    if endpoint:
        os.environ[FIXTURE_ENDPOINT_ENV] = str(endpoint)
    otel = str(request.get("otel_endpoint") or os.environ.get(OTEL_ENDPOINT_ENV) or "").rstrip("/")
    ingest = str(request.get("ingest_url") or os.environ.get(INGEST_URL_ENV) or "")
    if otel:
        os.environ.setdefault(OTEL_ENDPOINT_ENV, otel)
        os.environ.setdefault("OTEL_EXPORTER_OTLP_PROTOCOL", "http/protobuf")
        os.environ.setdefault("OTEL_EXPORTER_OTLP_TRACES_ENDPOINT", otel + "/v1/traces")
        if not ingest:
            ingest = otel + "/v1/runs"
    if ingest:
        os.environ.setdefault(INGEST_URL_ENV, ingest)
    if sample_id:
        existing = os.environ.get(RESOURCE_ATTRS_ENV, "")
        token = SAMPLE_ID_ATTR + "=" + sample_id
        if token not in existing.split(","):
            os.environ[RESOURCE_ATTRS_ENV] = (existing + "," + token).strip(",") if existing else token
    return sample_id


def _maybe_export(run: Dict[str, Any]) -> None:
    if os.environ.get(INGEST_URL_ENV):
        post_run(run)


def run_sample(handler: SampleHandler, stdin: Optional[IO[str]] = None, stdout: Optional[IO[str]] = None) -> Dict[str, Any]:
    """Run one sample for ``--runner exec``: stdin JSON in, AgentRun JSON out."""
    request = read_invoke(stdin)
    sample_id = apply_invoke_env(request)
    endpoint = request.get("tool_endpoint") or os.environ.get(FIXTURE_ENDPOINT_ENV) or None
    fixtures = FixtureClient(endpoint=endpoint)
    run = _as_dict(handler(request, fixtures), sample_id)
    _maybe_export(run)
    sink = stdout if stdout is not None else sys.stdout
    sink.write(json.dumps(run) + "\n")
    sink.flush()
    return run


def run_eval(handler: Callable[[Dict[str, Any]], Union[Dict[str, Any], RunRecorder, None]], stdin: Optional[IO[str]] = None, stdout: Optional[IO[str]] = None) -> Dict[str, Any]:
    """Minimal eval entrypoint: one function over scenario input/context.

    Use this when world control is ``existing`` (your own mocks/DI) and you either
    return an AgentRun / RunRecorder or rely on OTel already capturing the run.
    """

    def _wrap(request: Dict[str, Any], fixtures: FixtureClient) -> Union[Dict[str, Any], RunRecorder]:
        result = handler(request)
        if result is None:
            # Placeholder completed run when an external tracer owns observation.
            rec = RunRecorder(
                agent_name=os.environ.get("GUST_AGENT_NAME", "agent"),
                agent_version=os.environ.get("GUST_AGENT_VERSION", "0"),
                task_input=str(request.get("input") or ""),
            )
            rec.complete(output="traced-externally")
            return rec
        return result

    return run_sample(_wrap, stdin=stdin, stdout=stdout)


def execution_receipt(*, status: str = "completed", trace_id: str = "", run_id: str = "", run: Optional[Dict[str, Any]] = None, error: str = "") -> Dict[str, Any]:
    """Build a trigger-runner completion receipt for remote QA / harness flows."""
    out: Dict[str, Any] = {"status": status}
    if trace_id:
        out["trace_id"] = trace_id
    if run_id:
        out["run_id"] = run_id
    if error:
        out["error"] = error
    if run is not None:
        out["run"] = run
    return out


def serve_sample(handler: SampleHandler, host: str = "127.0.0.1", port: int = 8080) -> ThreadingHTTPServer:
    """Serve ``POST /invoke`` for ``--runner http``. Returns the started server.

    Call ``server.serve_forever()`` in your ``__main__`` block.
    """

    class Handler(BaseHTTPRequestHandler):
        def log_message(self, fmt: str, *args: Any) -> None:
            sys.stderr.write("%s - %s\n" % (self.address_string(), fmt % args))

        def do_POST(self) -> None:  # noqa: N802
            path = urlparse(self.path).path
            if path not in ("/invoke", "/"):
                self.send_error(404)
                return
            length = int(self.headers.get("Content-Length", "0"))
            raw = self.rfile.read(length) if length else b"{}"
            try:
                request = json.loads(raw.decode("utf-8") or "{}")
            except json.JSONDecodeError:
                self.send_error(400, "invalid JSON")
                return
            if not request.get("sample_id"):
                request["sample_id"] = self.headers.get("X-Gust-Sample-Id") or os.environ.get(SAMPLE_ID_ENV, "")
            if not request.get("traceparent"):
                request["traceparent"] = self.headers.get("traceparent") or ""
            if not request.get("baggage"):
                request["baggage"] = self.headers.get("baggage") or ""
            sample_id = apply_invoke_env(request)
            endpoint = request.get("tool_endpoint") or os.environ.get(FIXTURE_ENDPOINT_ENV) or None
            fixtures = FixtureClient(endpoint=endpoint)
            try:
                run = _as_dict(handler(request, fixtures), sample_id)
                _maybe_export(run)
            except Exception as exc:  # surface as a failed run, not a 500, when possible
                self.send_error(500, str(exc))
                return
            body = json.dumps(run).encode("utf-8")
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)

    server = ThreadingHTTPServer((host, port), Handler)
    return server

File: sdk/typescript/lib/sample.js
import http from "node:http";
import { FixtureClient, FIXTURE_ENDPOINT_ENV } from "./fixtures.js";
import { INGEST_URL_ENV, OTEL_ENDPOINT_ENV, RunRecorder, postRun } from "./recorder.js";
export const SAMPLE_ID_ENV = "AGENTEVAL_SAMPLE_ID";
const RESOURCE_ATTRS_ENV = "OTEL_RESOURCE_ATTRIBUTES";
const SAMPLE_ID_ATTR = "gust.sample_id";
export function applySampleId(run, sampleId) {
    if (!sampleId)
        return run;
    if (run instanceof RunRecorder) {
        run.runId = sampleId;
        run.setMetadata("sample_id", sampleId);
        return run;
    }
    run.run_id = sampleId;
    run.metadata = { ...(run.metadata || {}), sample_id: sampleId };
    return run;
}
function asDict(run, sampleId) {
    const stamped = applySampleId(run, sampleId);
    return stamped instanceof RunRecorder ? stamped.toDict() : stamped;
}
export function applyInvokeEnv(request) {
    const sampleId = String(request.sample_id || process.env[SAMPLE_ID_ENV] || "");
    if (sampleId)
        process.env[SAMPLE_ID_ENV] = sampleId;
    for (const [key, envName] of [
        ["evaluation_id", "GUST_EVALUATION_ID"],
        ["scenario_id", "GUST_SCENARIO_ID"],
        ["trace_id", "GUST_TRACE_ID"],
        ["traceparent", "TRACEPARENT"],
        ["baggage", "BAGGAGE"],
        ["world_control", "GUST_WORLD_CONTROL"],
    ]) {
        const value = request[key];
        if (value)
            process.env[envName] = String(value);
    }
    const endpoint = String(request.tool_endpoint || process.env[FIXTURE_ENDPOINT_ENV] || "");
    if (endpoint)
        process.env[FIXTURE_ENDPOINT_ENV] = endpoint;
    const otel = String(request.otel_endpoint || process.env[OTEL_ENDPOINT_ENV] || "").replace(/\/$/, "");
    let ingest = String(request.ingest_url || process.env[INGEST_URL_ENV] || "");
    if (otel) {
        if (!process.env[OTEL_ENDPOINT_ENV])
            process.env[OTEL_ENDPOINT_ENV] = otel;
        if (!process.env.OTEL_EXPORTER_OTLP_PROTOCOL)
            process.env.OTEL_EXPORTER_OTLP_PROTOCOL = "http/protobuf";
        if (!process.env.OTEL_EXPORTER_OTLP_TRACES_ENDPOINT) {
            process.env.OTEL_EXPORTER_OTLP_TRACES_ENDPOINT = otel + "/v1/traces";
        }
        if (!ingest)
            ingest = otel + "/v1/runs";
    }
    if (ingest && !process.env[INGEST_URL_ENV])
        process.env[INGEST_URL_ENV] = ingest;
    if (sampleId) {
        const existing = process.env[RESOURCE_ATTRS_ENV] || "";
        const token = `${SAMPLE_ID_ATTR}=${sampleId}`;
        if (!existing.split(",").includes(token)) {
            process.env[RESOURCE_ATTRS_ENV] = existing ? `${existing},${token}` : token;
        }
    }
    return sampleId;
}
export function sampleContext() {
    return {
        evaluation_id: process.env.GUST_EVALUATION_ID || "",
        scenario_id: process.env.GUST_SCENARIO_ID || "",
        sample_id: process.env[SAMPLE_ID_ENV] || "",
        trace_id: process.env.GUST_TRACE_ID || "",
        traceparent: process.env.TRACEPARENT || "",
        baggage: process.env.BAGGAGE || "",
        world_control: process.env.GUST_WORLD_CONTROL || "",
        tool_endpoint: process.env[FIXTURE_ENDPOINT_ENV] || "",
        input: process.env.AGENTEVAL_TASK_INPUT || "",
    };
}
export function executionReceipt({ status = "completed", trace_id = "", run_id = "", run, error = "" } = {}) {
    const out = { status };
    if (trace_id)
        out.trace_id = trace_id;
    if (run_id)
        out.run_id = run_id;
    if (error)
        out.error = error;
    if (run)
        out.run = run;
    return out;
}
async function maybeExport(run) {
    if (process.env[INGEST_URL_ENV]) {
        await postRun(run);
    }
}
export async function runSample(handler, { input, output } = {}) {
    const chunks = [];
    const stream = input || process.stdin;
    if (stream !== process.stdin || !stream.isTTY) {
        for await (const chunk of stream)
            chunks.push(Buffer.from(chunk));
    }
    const raw = Buffer.concat(chunks).toString("utf8").trim();
    const request = raw ? JSON.parse(raw) : {};
    const sampleId = applyInvokeEnv(request);
    const endpoint = String(request.tool_endpoint || process.env[FIXTURE_ENDPOINT_ENV] || "");
    const fixtures = new FixtureClient({ endpoint });
    const run = asDict(await handler(request, fixtures), sampleId);
    await maybeExport(run);
    const sink = output || process.stdout;
    sink.write(JSON.stringify(run) + "\n");
    return run;
}
/** Minimal eval entrypoint when world_control is existing (your mocks/DI). */
export async function runEval(handler, opts) {
    return runSample(async (request) => {
        const result = await handler(request);
        if (result == null) {
            const rec = new RunRecorder({
                agentName: process.env.GUST_AGENT_NAME || "agent",
                agentVersion: process.env.GUST_AGENT_VERSION || "0",
                taskInput: String(request.input || ""),
            });
            rec.complete("traced-externally");
            return rec;
        }
        return result;
    }, opts);
}
export function serveSample(handler, { host = "127.0.0.1", port = 8080 } = {}) {
    const server = http.createServer(async (req, res) => {
        if (req.method !== "POST" || (req.url !== "/invoke" && req.url !== "/")) {
            res.writeHead(404);
            res.end();
            return;
        }
        const chunks = [];
        for await (const chunk of req)
            chunks.push(Buffer.from(chunk));
        let request = {};
        try {
            request = JSON.parse(Buffer.concat(chunks).toString("utf8") || "{}");
        }
        catch {
            res.writeHead(400);
            res.end("invalid JSON");
            return;
        }
        if (!request.sample_id) {
            request.sample_id = String(req.headers["x-gust-sample-id"] || process.env[SAMPLE_ID_ENV] || "");
        }
        const sampleId = applyInvokeEnv(request);
        const endpoint = String(request.tool_endpoint || process.env[FIXTURE_ENDPOINT_ENV] || "");
        const fixtures = new FixtureClient({ endpoint });
        try {
            const run = asDict(await handler(request, fixtures), sampleId);
            await maybeExport(run);
            const body = JSON.stringify(run);
            res.writeHead(200, { "Content-Type": "application/json", "Content-Length": Buffer.byteLength(body) });
            res.end(body);
        }
        catch (err) {
            res.writeHead(500);
            res.end(String(err instanceof Error ? err.message : err));
        }
    });
    server.listen(port, host);
    return server;
}
